### PR TITLE
[SPARK-11899][SQL] API audit for GroupedDataset.

### PR DESCRIPTION
**Pull request by rxin**
1. Renamed map to mapGroup, flatMap to flatMapGroup.
2. Renamed asKey -> keyAs.
3. Added more documentation.
4. Changed type parameter T to V on GroupedDataset.
5. Added since versions for all functions.

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala [hunk 1]
```diff
@@ -36,11 +37,13 @@ import org.apache.spark.sql.execution.QueryExecution
...
  * after Spark 1.6.
+ *
+ * @since 1.6.0
  */
 @Experimental
-class GroupedDataset[K, T] private[sql](
+class GroupedDataset[K, V] private[sql](
     kEncoder: Encoder[K],
-    tEncoder: Encoder[T],
+    tEncoder: Encoder[V],
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: Should this variable be renamed to `vEncoder`?

**Comment by JoshRosen** [comment]
LGTM aside from one minor variable naming nit.

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: yea that's a good catch. will fix it in my next pr.

**Pull request closed without merging**